### ISSUE TITLE
r39-5 log issue

### ISSUE THREAD

**Issue by stabker** [pBlueG/SA-MP-MySQL#91]
I have next code

```
new i;
for(i = 0; i < MAX_FARMS; i++) SaveFarm(i);
for(i = 0; i < TOTAL_HOUSES; i++) SaveHouse(i);
mysql_close(connection);
//mysql_close waiting until all queued queries are executed
SendRconCommand("exit");
```

And after upgrading to r39-5 next message started prints to mysql_log:

> [WARNING] CMySQLConnection::Disconnect - no connection available

Sorry for my bad English.

**Comment by maddinat0r** [comment]
Does this happen every time? Can you reproduce that with debug logging enabled?

**Comment by Kar2k** [comment]
I must say I usually get that alot too on windows. But it doesn't really bother me. 50/50 it says that.

**Comment by stabker** [comment]
Looks like it happens in most cases, but i'm not sure that it happens every time. Here is mysql_log in LOG_ERROR | LOG_WARNING | LOG_DEBUG log-level:

http://pastebin.com/m0xn2CB0

**Comment by IstuntmanI** [comment]
Where are you using that code ? If that code is executed in a command, you are closing the connection in that command, then it is "exiting" the server, so it will be calling OnGameModeExit where I guess you have another mysql_close called, but because you already closed the connection, that warning is getting printed.

I have no problems with this. I use R39-5 since it was released and I have no weird errors/warnings.

**Comment by Kar2k** [comment]
I get this issue sometimes. I have one ongamemodeexit. Unless a plugin is calling ongamemodeexit.. something is wrong. 

```pawn
public OnGameModeExit()
{ // The result is already empty
	if(gmxDisableCalls > 0) return 0;
	new uptime_seconds, days, hours, minutes, seconds, remaining;
	uptime_seconds = (gettime() - ServerInfo[Time_S_Started]);
	remaining = uptime_seconds % 3600; // remaining seconds after subtracting hours
	hours = (uptime_seconds - remaining) / 3600;
	seconds = remaining % 60;
	minutes = (remaining - seconds) / 60;
	while(hours >= 24)
	{
		hours -= 24;
		++days;
	}
	KillTimer(RealTimeUpdate_Timer);
	KillTimer(Online_Map_Timer);
	for(new cityID = LOS_SANTOS; cityID < MAX_CITIES; ++cityID) {
		if(iFireTimer[cityID] == -1) continue;
		KillTimer(iFireTimer[cityID]);
	}
	for(new t = 0; t < MAX_TEXT_DRAWS; t++) {
		TextDrawDestroy(Text:t);
	}
	for(new i = 0; i < MAX_OBJECTS; i++) {
		DestroyObject(i);
	}
	for(new i = 0; i < MAX_3DTEXT_GLOBAL; i++) {
		Delete3DTextLabel(Text3D:i);
	}
	destroyAllSpikeStrips();
	#if SAMP_037 == (true)
	destroyActors();
	#endif
	#if defined socket_included
	destroyServerSockets();
	#endif
	MapAndreas_Unload();
	quitIRCBots(true);
	
	#if defined _regex_included
	regex_delete_all();
	#endif

	Streamer_DestroyAllItems(STREAMER_TYPE_OBJECT);
	Streamer_DestroyAllItems(STREAMER_TYPE_PICKUP);
	Streamer_DestroyAllItems(STREAMER_TYPE_CP);
	Streamer_DestroyAllItems(STREAMER_TYPE_MAP_ICON);
	if(MYSQL_CONNECTED_STATUS == true) { // HERE
		mysql_close(mysqlConnHandle);
	}
	#if defined tsconnector_included
	disconnectTeamspeak();
	#endif
	printf("[OnGameModeExit:: Heapspace] - heapspace(): %i - Heapspace: %i kilobytes.", heapspace(), heapspace() / 1024);
	printf("[SYSTEM: OnGameModeExit] - uptime: %d - %d days, %d hours, %d minutes, %d seconds.", uptime_seconds, days, hours, minutes, seconds);
	return 1;
}
```

`[15:25:57] [WARNING] CMySQLConnection::Disconnect - no connection available

**Comment by maddinat0r** [comment]
Are you using `mysql_pquery`? If so, to which value do you set the `pool_size` parameter in `mysql_connect`? How many `connection was closed` debug messages do you get before and after that warning?

**Comment by stabker** [comment]
IstuntmanI, nope, i have just one mysql_close and it located in code, which i showed earlier.

maddinat0r, if the question to me, i'm only use a mysql_tquery. pool_size is default (2). As you can see here http://pastebin.com/daEhsfKE i have 3 "connection was closed" messages before this warning.

**Comment by Kar2k** [comment]
for me, pool size 0. one disconnect message. "CMySQLConnection::Disconnect - no connection available" just that alone.

Windows 10

**Comment by maddinat0r** [comment]
I *think* I found the issue. It's really nothing but an useless warning anyway, so you can safely ignore it.
If any of you want a build where it's fixed, leave a comment here (with the OS you need it for).

**Comment by Kar2k** [comment]
```
[18:32:30] [WARNING] CMySQLConnection::Disconnect - no connection available
[21:14:42] [WARNING] CMySQLConnection::Disconnect - no connection available
[21:17:16] [WARNING] CMySQLConnection::Disconnect - no connection available
[21:21:14] [WARNING] CMySQLConnection::Disconnect - no connection available
[21:34:33] [WARNING] CMySQLConnection::Disconnect - no connection available
[21:40:43] [WARNING] CMySQLConnection::Disconnect - no connection available
[21:41:57] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:09:43] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:12:56] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:25:32] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:32:26] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:39:24] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:48:35] [WARNING] CMySQLConnection::Disconnect - no connection available
[22:51:09] [WARNING] CMySQLConnection::Disconnect - no connection available
[23:00:14] [WARNING] CMySQLConnection::Disconnect - no connection available
[01:52:26] [WARNING] CMySQLConnection::Disconnect - no connection available
[02:17:03] [WARNING] CMySQLConnection::Disconnect - no connection available
[02:26:46] [WARNING] CMySQLConnection::Disconnect - no connection available
[00:20:50] [WARNING] CMySQLConnection::Disconnect - no connection available
```

All in 1 night of testing (starting, closing, gmxing etc).

Sure, erm windows 10.

It's not a big issue, but the less bugs the merrier :palm_tree:

**Comment by maddinat0r** [comment]
Here you go: http://en.file-upload.net/download-11512726/mysql-R39-5-4-gb32cef1-win32.zip.html

**Comment by Kar2k** [comment]
Still happening

**Comment by WopsS** [comment]
Hello,

I did a debugging for this issue and this is what I got:

I will call **mysql_close** from PAWN at **OnGameModeExit** and it will call this **Handle->ExecuteOnConnections(boost::bind(&CMySQLConnection:: Disconnect, _1));**, that's OK because it will queue **CMySQLConnection:: Disconnect** in **m_FuncQueue** _(**the problem is just for threaded connections** but I can't reproduce it with **mysql_pquery**)_, but right after **Handle->ExecuteOnConnections(boost::bind(&CMySQLConnection:: Disconnect, _1));** finish it's execution **Handle->Destroy();** is called, that's OK too but here is the problem! 

OK, let's continue in **Handle->Destroy();** it will call delete on the class instance
```cpp
if (ActiveHandle == this)
	ActiveHandle = NULL;

SQLHandle.erase(m_MyID);
delete this;
```

now in **CMySQLHandle::~CMySQLHandle** is called **ExecuteOnConnections(boost::bind(&CMySQLConnection:: Destroy, _1));** which will call again
```cpp
if (m_ThreadConnection != NULL)
	func(m_ThreadConnection);
	
for(set<CMySQLConnection *>::iterator c = m_ConnectionPool.begin(), end = m_ConnectionPool.end(); c != end; ++c)
	func(*c);
```

but with another function **CMySQLConnection:: Destroy** now here is the problem _(since the unthreaded connection doesn't need to use **CMySQLConnection::ProcessQueries** to close the connection it won't affect it)_ because **CMySQLConnection::ProcessQueries** doesn't have time to close the connection properly and set **m_IsConnected** to **false** it will requeue the **CMySQLConnection:: Disconnect**, so now there are 2 **CMySQLConnection:: Disconnect** in **m_FuncQueue**.

You can fix it with this code in **CMySQLConnection:: Destroy**
```cpp
if (m_QueryThread != NULL)
{
	bool func_queue_empty = false;
	do
	{
		m_FuncQueueMtx.lock();
		func_queue_empty = m_FuncQueue.empty();
		m_FuncQueueMtx.unlock();
		this_thread::sleep_for(chrono::milliseconds(5));
	} while (func_queue_empty == false);
}

if(m_IsConnected)
	Disconnect();
delete this;
```

Testing code:
```pawn
#include <a_samp>
#include <a_mysql>

new ConnectionHandle;

forward CloseServer();

main()
{
}

public OnGameModeInit()
{
	mysql_log(LOG_ALL, LOG_TYPE_HTML);
	
	ConnectionHandle = mysql_connect("localhost", "root", "samp", "", 3306, true, 1);
	SetTimer("CloseServer", 3000, false);
	
	return 1;
}

public OnGameModeExit()
{
	mysql_tquery(ConnectionHandle, "SHOW TABLES");
	mysql_tquery(ConnectionHandle, "SHOW TABLES");
	
 	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	mysql_pquery(ConnectionHandle, "SHOW TABLES");
	
	mysql_close(ConnectionHandle);
	
	return 1;
}

public CloseServer()
{
    SendRconCommand("exit");
}
```

**Comment by maddinat0r** [comment]
Thanks for your debugging @WopsS! You are right, the queued function calls are not processed fast enough, thus leading to two disconnect calls for the same connection in the queue.

I've applied a much simpler fix though (2186e72bc92165eecca23b61e5861feb5286908b) and actually tested and debugged everything by myself now, so at this point I can actually guarantee that this bug is now fixed for good.
